File: VS/lab3/Unit.cpp
#include "Unit.hpp"
#include "Staff.hpp"
#include "SupplyRequestDetail.hpp"
#include "SupplyResponse.hpp"
#include <algorithm>

Unit::Unit(int id, const std::string& name, UnitType unitType) :
    m_id(id), m_name(name), m_type(unitType) {
}

UnitType Unit::GetType() const {
    return m_type;
}

void Unit::SetId(int id) {
    m_id = id;
}
int Unit::GetId() const {
    return m_id;
}

std::string Unit::GetName() const {
    return m_name;
}

void Unit::SetParent(Unit* parent) {
    if (parent) {
        m_parentId = parent->m_id;
        m_parent = parent;
    }
}

SupplyResponse Unit::MakeSupplyRequest(SupplyRequest& request) {
    bool detailsCorreted = m_assignedWarehouse != nullptr ?
        m_assignedWarehouse->ProcessSupplyRequestDetails(request.GetDetails()) : false;

    auto& remainingDetails = request.GetDetails();

    if (remainingDetails.empty())
        return SupplyResponse(SupplyResponseStatus::Success, "");


    if (m_parent != nullptr) {
        auto parentResponse = m_parent->MakeSupplyRequest(request);

        if (parentResponse.GetStatus() == SupplyResponseStatus::Success)
            return SupplyResponse(SupplyResponseStatus::Success, "");

        return detailsCorreted ? SupplyResponse(SupplyResponseStatus::Partial, SupplyRequestDetail::ToString(remainingDetails)) :
            SupplyResponse(SupplyResponseStatus::Denied, "Ни один из складов высших порядков не смог удовлетворить запрос");
    }

    return detailsCorreted ? SupplyResponse(SupplyResponseStatus::Partial, SupplyRequestDetail::ToString(remainingDetails)) :
        SupplyResponse(SupplyResponseStatus::Denied, m_assignedWarehouse == nullptr ? "Нет прикрепленного склада" : "Склад подразделения не смог удовлетворить запрос");
}

void Unit::AddChildUnit(Unit* unit) {
    if (unit && (unit->GetType() < m_type)) {
        unit->SetParent(this);
        m_children.push_back(unit);
    }
}

void Unit::AddChildUnits(std::vector<Unit*>& units) {
    for (Unit* unit : units)
        m_children.push_back(unit);
}

void Unit::AddSoldier(Staff* soldier) {
    m_personnel.push_back(soldier);
}
void Unit::AddSoldiers(std::vector<Staff*>& soldiers) {
    for (auto& soldier : soldiers)
        m_personnel.push_back(soldier);
}
bool Unit::RemoveSoldier(int id) {
    int startSize = m_personnel.size();
    auto newEnd = std::remove_if(m_personnel.begin(), m_personnel.end(), [&id](const Staff* soldier)
        {
            return soldier == nullptr || soldier->GetId() == id;
        });
    m_personnel.erase(newEnd, m_personnel.end());
    return m_personnel.size() < startSize;
}

void Unit::AssignWarehouse(Warehouse* warehouse) {
    if (warehouse) {
        m_assignedWarehouseId = warehouse->GetId();
        m_assignedWarehouse = warehouse;
    }
}

void Unit::AssignCommander(const Staff* commander) {
    if (commander) {
        m_commanderId = commander->GetId();
        m_commander = commander;
    }
}

bool Unit::RemoveChildUnit(int id) {

    int prevSize = m_children.size();

    auto newEnd = std::remove_if(m_children.begin(), m_children.end(), [&id](const Unit* unit)
        {
            if (unit)
                return unit->GetId() == id;

            return true;
        });

    m_children.erase(newEnd, m_children.end());

    return m_children.size() <= prevSize;
}

SupplyRequest Unit::CreateRequest(std::vector<std::unique_ptr<SupplyRequestDetail>>& details) const {
    return SupplyRequest(rand(), this, std::move(details));
}

File: lab8/Java/src/unit/Unit.java
package unit;

import enums.SupplyResponseStatus;
import enums.UnitType;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import request.SupplyRequest;
import request.SupplyRequestDetail;
import request.SupplyResponse;
import warehouse.Warehouse;

public class Unit implements Suppliable {

    int id;
    String name;
    UnitType type;
    int parentId;
    Unit parent;
    ArrayList<Unit> children = new ArrayList<>();
    ArrayList<Staff> personnel = new ArrayList<>();
    int commanderId;
    Staff commander;
    int assignedWarehouseId;
    Warehouse assignedWarehouse;

    public Unit(int id, String name, UnitType type){
        this.id = id;
        this.name = name;
        this.type = type;
    }

    public int getId(){
        return id;
    }

    public String getName(){
        return name;
    }

    public UnitType getType(){
        return type;
    }

    public Unit getParent(){
        return parent;
    }

    public Staff getCommander(){
        return commander;
    }

    public void setParent(Unit parent){
        if(parent != null && parent.type.compareTo(type) > 0){
            this.parentId = parent.getId();
            this.parent = parent;
        }
    }

    public void addChildUnit(Unit unit){
        if(unit != null && !hasChild(unit) && unit.getType().compareTo(type) < 0){
            this.children.add(unit);
            unit.setParent(this);
        }
    }

    public int addChildUnits(ArrayList<Unit> units){
        var copy = new ArrayList<Unit>(units);
        copy.removeIf(i->i == null || i.getType().compareTo(type) > 0);
        copy.removeIf(i->children.stream().anyMatch(j -> j.getId() == i.getId()));
        var distincted = copy.stream().filter(distinctBy(u->u.getId())).collect(Collectors.toList());
        for(Unit unit : distincted){
            unit.setParent(this);
        }
        children.addAll(distincted);
        return distincted.size();
    }

    public Boolean removeChildUnit(int id){
        return children.removeIf(c->c.getId() == id);
    }

    public List<Unit> getChildren(){
        return Collections.unmodifiableList(children);
    }    

    public void addSoldier(Staff soldier){
        if(soldier != null && !hasPersonnel(soldier)){
            soldier.setUnit(this);
            personnel.add(soldier);
        }
    }

    public int addSoldiers(ArrayList<Staff> soldiers){
        var copy = new ArrayList<Staff>(soldiers);
        copy.removeIf(i->personnel.stream().anyMatch(j->j.getId() == i.getId()));
        var distincted = copy.stream().filter(distinctBy(s->s.getId())).collect(Collectors.toList());
        for(Staff soldier : distincted){
            soldier.setUnit(this);
        }        
        personnel.addAll(distincted);
        return distincted.size();
    }

    public Boolean removeSoldier(int id){
        return personnel.removeIf(p->p.getId() == id);
    }

    public List<Staff> getPersonnel(){
        return Collections.unmodifiableList(personnel);
    }

    public ArrayList<Staff> findByName(String name){
        ArrayList<Staff> result = new ArrayList<>();
        for(var soldier : personnel){
            if(soldier.fullName.toLowerCase().contains(name)){
                result.add(soldier);
            }
        }
        return result;
    }

    public void assignWarehouse(Warehouse warehouse){
        if(warehouse != null){
            this.assignedWarehouseId = warehouse.getId();
            this.assignedWarehouse = warehouse;
        }
    }

    public void assignCommander(Staff commander){
        // добавить в список staff
        if(commander != null){
            this.commanderId = commander.getId();
            this.commander = commander;
        }
    }

    @Override
    public SupplyResponse makeSupplyRequest(SupplyRequest request){

        Boolean isCorrected = false;

        if(request == null)
            return new SupplyResponse(SupplyResponseStatus.DENIED, "Передан пустой запрос"); 

        
        if(assignedWarehouse != null){
            isCorrected = assignedWarehouse.processSupplyRequestDetails(request.getDetails());
        }

        if(!request.getDetails().isEmpty()){
            if(parent != null){
                var parentResponse = parent.makeSupplyRequest(request);
                
                if(parentResponse.getStatus() == SupplyResponseStatus.DENIED)
                    return isCorrected ? new SupplyResponse(SupplyResponseStatus.PARTIAL, request.getDetails()) :  parentResponse;

                return parentResponse;

            }
            return new SupplyResponse(SupplyResponseStatus.DENIED, "Ни один из складов дерева не смог удовлетворить запрос");

        }

        return new SupplyResponse(SupplyResponseStatus.SUCCESS, "");
    }

    public List<String> getStafflList(){
        return personnel.stream().map(p->p.toString()).toList();
    }
    public List<Staff> findStaffByFullName(String fullName){
        return personnel.stream().filter(p->p.getFullName().toLowerCase().contains(fullName.toLowerCase())).toList();
    }

    public SupplyRequest createRequest(ArrayList<SupplyRequestDetail> details){
        return new SupplyRequest(0, details, this);
    }

    Boolean hasPersonnel(Staff soldier){
        return personnel.stream().anyMatch(s->s.getId() == soldier.getId());
    }

    Boolean hasChild(Unit child){
        return children.stream().anyMatch(s->s.getId() == child.getId());
    }    

    static <T> Predicate<T> distinctBy(Function<? super T, ?> keyExtractor) {
        Set<Object> seen = ConcurrentHashMap.newKeySet();
        return t -> seen.add(keyExtractor.apply(t));
    }
}
